### PR TITLE
[IMP] mail: chat window hidden

### PR DESCRIPTION
**Pull request by phenix-factory**

### CODE REVIEW

**Comment by alexkuhn** [comment]
👆rebased (target branch was rebased to master)

File: addons/mail/static/src/new/chat/chat_window_hidden_menu.scss [hunk 1]
```diff
@@ -0,0 +1,12 @@
+.o-mail-chat-window-hidden-menu {
+    z-index: $zindex-dropdown;
```
Comment (seb-odoo) [inline addons/mail/static/src/new/chat/chat_window_hidden_menu.scss]: strange to defile this manually when using Dropdown component?

File: addons/mail/static/src/new/chat/chat_window_container.xml [hunk 2]
```diff
@@ -2,8 +2,11 @@
 <templates xml:space="preserve">
 
 <t t-name="mail.chat_window_container" owl="1">
-    <t t-foreach="chatWindows" t-as="chatWindow" t-key="chatWindow.threadLocalId">
-        <ChatWindow chatWindow="chatWindow" right="10 + chatWindow_index * 350"/>
+    <t t-if="!messaging.state.discuss.isActive">
+        <ChatWindowHiddenMenu t-if="messaging.hiddenChatWindows.length > 0" chatWindows="messaging.hiddenChatWindows" />
+        <t t-foreach="messaging.visibleChatWindows" t-as="chatWindow" t-key="chatWindow.threadLocalId">
+            <ChatWindow chatWindow="chatWindow" right="10 + chatWindow_index * 350"/>
```
Comment (seb-odoo) [inline addons/mail/static/src/new/chat/chat_window_container.xml]: compute of "right" would be better with const defined in other file

File: addons/mail/static/src/new/core/chat_window_model.js [hunk 3]
```diff
@@ -2,13 +2,48 @@
...
+
 export class ChatWindow {
-    /** @type {number} */
-    autofocus = 0;
-    /** @type {boolean} */
+    /** @type {import("@mail/new/core/messaging").Messaging['state']} */
+    _state;
+
+    /** @type {string|number} */
+    threadLocalId;
```
Comment (seb-odoo) [inline addons/mail/static/src/new/core/chat_window_model.js]: can only be `string` now

File: addons/mail/static/src/new/core/chat_window_model.js [hunk 4]
```diff
@@ -2,13 +2,48 @@
...
+    autofocus = 1;
```
Comment (seb-odoo) [inline addons/mail/static/src/new/core/chat_window_model.js]: looks really bad to have an autofocus by default, I would not change that

File: addons/mail/static/src/new/core/chat_window_model.js [hunk 5]
```diff
@@ -35,6 +70,12 @@ export class ChatWindow {
             _state: state,
         });
         this.update(data);
+        if (ChatWindow.maxVisibleChatWindows(this._state) <= this._state.chatWindows.length) {
+            const visibleChatWindows = ChatWindow.visibleChatWindows(this._state);
+            const swaped = visibleChatWindows[visibleChatWindows.length - 1];
+            swaped.hidden = true;
+            swaped.folded = true;
```
Comment (seb-odoo) [inline addons/mail/static/src/new/core/chat_window_model.js:77]: strange to sync folded with hidden.

Also other methods that do show/hide call fold methods which also update fold state on thread, whereas from here the thread is not updated according to change in chat window.

File: addons/mail/static/src/new/core/chat_window_model.js [hunk 6]
```diff
@@ -81,11 +123,42 @@ export class ChatWindow {
...
+        this.hidden = true;
+        this.fold();
+    }
+
+    show() {
+        this.hidden = false;
+        this.unfold();
+    }
+
+    swapHidden() {
```
Comment (seb-odoo) [inline addons/mail/static/src/new/core/chat_window_model.js]: name of this method is really not clear

File: addons/mail/static/src/new/core/messaging.js [hunk 7]
```diff
@@ -94,7 +94,11 @@ export class Messaging {
             });
         });
         this.registeredImStatusPartners = reactive([], () => this.updateImStatusRegistration());
+        const self = this;
         this.state = reactive({
+            get isSmall() {
+                return self.env.isSmall;
```
Comment (seb-odoo) [inline addons/mail/static/src/new/core/messaging.js]: could do env.isSmall without storing a self

File: addons/mail/static/src/new/core/chat_window_model.js (hunk 6 above)
Comment (alexkuhn) [reply addons/mail/static/src/new/core/chat_window_model.js]: renamed to `makeVisible`

**Review by seb-odoo — COMMENTED** [review]
see comments, everything is not perfect but ok for a first pass

**Pull request merged**